lib/internal/abort_controller.js

```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+// Modeled very closely on the AbortController implementation
+// in https://github.com/mysticatea/abort-controller (MIT license)
+
+const {
+  SafeWeakMap,
+  Object,
+  Symbol,
+} = primordials;
+const EventEmitter = require('events');
+const {
+  codes: { ERR_INVALID_THIS }
+} = require('internal/errors');
+
+const signals = new SafeWeakMap();
+const controllers = new SafeWeakMap();
+
+class AbortSignal extends EventEmitter {
```

Comment:
I was wondering how hard it would be to pull a mostly compatible EventTarget like Deno's (I spoke to Kitson a bit) since they are [quite different](https://docs.google.com/document/d/1NFARs04-4U_2y6Ssw9Lqu1GMXBUM981-NO9PLJWifTI/edit?usp=sharing) (as you saw).

How do you feel about a subclass of EventEmitter that implements some (most?) of the EventTarget stuff vs. just pulling EventTarget from Deno/another shim vs. just having AbortSignal implement EE and not ET?

Reply:
> How do you feel about a subclass of EventEmitter that implements some (most?) of the EventTarget stuff ...

Already exploring it ;-)

Reply:
A big point is how our shim can be parent-less - so we can skip the getParent part Deno implemented and are probably removing.

That is, we don't need phases, preventDefault, stuff related to shadow dom etc. If it would help I don't mind making a gist of what such an implementation might look like (either based on EventEmitter or independently)

Reply:
Yep. The biggest complexity we have on it really is making it somewhat compatible with EventEmitter

Reply:
I was thinking that the APIs that take AbortSignal ought to be agnostic to whether the AbortSignal was a EventTarget AbortSignal, node-provided AbortSignal object, or something else. So technically, we'd only need to implement the EventTarget features that were actually used in cancellation mechanisms, which so far in my branch is just addEventListener and removeEventListener.

Reply:
There was a plan to standardize the interface here: https://github.com/tc39/proposal-cancellation

Reply:
See #33556